### PR TITLE
feat: grinding

### PR DESCRIPTION
**Pull request by yagnadeepxo**

### CODE REVIEW

**Comment by yagnadeepxo** [comment]
[dependencies]
sha3 = "0.10.8"

[[test]]
name = "grinding_test"
path = "src/grinding.rs"


cargo.toml

File: src/grinding.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    let computed_hash = hash(seed_data, pow_bits);
+    let threshold = 1 << (64 - pow_bits);
+    check_nonce_validity(&computed_hash, trial_nonce, threshold)
+}
+
+// Attempts to find a valid nonce that satisfies the grinding criteria for a given seed and difficulty level.
+pub fn find_valid_nonce(seed_data: &[u8; 32], difficulty_level: u8) -> Option<u64> {
+    let computed_hash = hash(seed_data, difficulty_level);
+    let threshold = 1 << (64 - difficulty_level);
+    (0..u64::MAX).find(|&test_nonce| {
```
Comment (utsavjnn) [inline src/grinding.rs]: @yagnadeepxo can you do it parallely? Not the best idea to go from `0 to u64::MAX` sequentially.

**Review by utsavjnn — CHANGES_REQUESTED** [review]

**Comment by utsavjnn** [comment]
Also would be great if you could make it part of the fri config, and integrate it in the actuall proof generation and verification process.

File: src/grinding.rs (hunk 1 above)
Comment (yagnadeepxo) [reply src/grinding.rs]: added with rayon

**Comment by yagnadeepxo** [comment]
added grinding to FRIConfig @utsavjnn

**Pull request closed without merging**